alembic/autogenerate/rewriter.py

```diff
@@ -23,41 +23,40 @@
     from ..runtime.environment import _GetRevArg
     from ..runtime.migration import MigrationContext
 
+    ProcessRevisionDirectiveFn = Callable[
+        [MigrationContext, _GetRevArg, List["MigrationScript"]], None
+    ]
 
-class Rewriter:
-    """A helper object that allows easy 'rewriting' of ops streams.
-
-    The :class:`.Rewriter` object is intended to be passed along
-    to the
-    :paramref:`.EnvironmentContext.configure.process_revision_directives`
-    parameter in an ``env.py`` script.    Once constructed, any number
-    of "rewrites" functions can be associated with it, which will be given
-    the opportunity to modify the structure without having to have explicit
-    knowledge of the overall structure.
-
-    The function is passed the :class:`.MigrationContext` object and
-    ``revision`` tuple that are passed to the  :paramref:`.Environment
-    Context.configure.process_revision_directives` function normally,
-    and the third argument is an individual directive of the type
-    noted in the decorator.  The function has the choice of  returning
-    a single op directive, which normally can be the directive that
-    was actually passed, or a new directive to replace it, or a list
-    of zero or more directives to replace it.
 
-    .. seealso::
+class RewriterBase:
+    """A helper object that makes `process_revision_directives` chainable.
 
-        :ref:`autogen_rewriter` - usage example
+    The :class:`.RewriterBase` object wraps a callable, so that in addition
+    to be passed along directly to the
+    :paramref:`.EnvironmentContext.configure.process_revision_directives`
+    parameter in an ``env.py`` script, it could be chained to other instances
+    of :class:`.RewriterBase` or its subclasses.
 
     """
 
-    _traverse = util.Dispatcher()
-
-    _chained: Optional[Rewriter] = None
+    _chained: Optional[RewriterBase] = None
 
-    def __init__(self) -> None:
-        self.dispatch = util.Dispatcher()
+    def __init__(
+        self,
+        process_revision_directives: Optional[
+            ProcessRevisionDirectiveFn
+        ] = None,
```

Comment:
**Federico Caselli** (CaselIT) wrote:

this could be required, meaning remove the `Optional` and the `=None`

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/4990

Reply:
I believe this no longer applies.